File: SpEditText/src/main/java/com/sunhapper/x/spedit/view/SpXSpannableFactory.java
package com.sunhapper.x.spedit.view;

import android.support.annotation.NonNull;
import android.text.NoCopySpan;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;

import java.util.List;

/**
 * Created by sunhapper on 2019/1/25 .
 */
public class SpXSpannableFactory extends Spannable.Factory {
    private List<NoCopySpan> mNoCopySpans;
    public SpXSpannableFactory(List<NoCopySpan> watchers) {
        mNoCopySpans = watchers;
    }

    public Spannable newSpannable(@NonNull CharSequence source) {
        SpannableStringBuilder spannableStringBuilder = new  SpannableStringBuilder();
        for (NoCopySpan span : mNoCopySpans) {
            spannableStringBuilder.setSpan(span, 0,0,
                    Spanned.SPAN_INCLUSIVE_INCLUSIVE | Spanned.SPAN_PRIORITY);
        }
        spannableStringBuilder.append(source);
        return spannableStringBuilder;
    }
}

File: SpEditText/src/main/java/com/sunhapper/x/spedit/view/SpXSpannableFactory.kt
package com.sunhapper.x.spedit.view

import android.text.NoCopySpan
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.Spanned

/**
 * Created by sunhapper on 2019/1/25 .
 */
class SpXSpannableFactory(private val mNoCopySpans: List<NoCopySpan>) : Spannable.Factory() {

    override fun newSpannable(source: CharSequence): Spannable {
        val spannableStringBuilder = SpannableStringBuilder()
        mNoCopySpans.forEach {
            spannableStringBuilder.setSpan(it, 0, 0,
                    Spanned.SPAN_INCLUSIVE_INCLUSIVE or Spanned.SPAN_PRIORITY)
        }
        spannableStringBuilder.append(source)
        return spannableStringBuilder
    }
}
